src/dmd/detect_unused.d

```diff
@@ -0,0 +1,275 @@
+        if (!sd.isReferenced &&
+            sd.protection.kind != Prot.Kind.undefined && // has semantics
+            sd.protection.kind.shouldWarnUnunsed())
+            sd.loc.warning("unused %s struct `%s`", sd.prot.kind.toChars(), sd.toChars());
+        super.visit(sd);
+    }
+
+    override void visit(ClassDeclaration cd)
+    {
+        assert(cd);
+        if (!cd.isReferenced &&
+            cd.protection.kind != Prot.Kind.undefined && // has semantics
+            cd.protection.kind.shouldWarnUnunsed())
+            cd.loc.warning("unused %s class `%s`", cd.prot.kind.toChars(), cd.toChars());
+        super.visit(cd);
+    }
+
+    override void visit(InterfaceDeclaration id)
+    {
+        assert(id);
+        if (!id.isReferenced &&
+            id.protection.kind != Prot.Kind.undefined && // has semantics
+            id.protection.kind.shouldWarnUnunsed())
+            id.loc.warning("unused %s interface `%s`", id.prot.kind.toChars(), id.toChars());
+        super.visit(id);
+    }
+
+    override void visit(TemplateDeclaration td)
+    {
+        assert(td);
+        if (!td.isReferenced &&
+            td.protection.kind != Prot.Kind.undefined && // has semantics
+            td.protection.kind.shouldWarnUnunsed())
+            td.loc.warning("unused %s template `%s`", td.prot.kind.toChars(), td.toChars());
+        super.visit(td);
+    }
+
+    override void visit(LabelDsymbol ls)
+    {
+        assert(ls);
+        if (!ls.isReferenced)
+            ls.loc.warning("unused %s label `%s`", ls.prot.kind.toChars(), ls.toChars());
+        super.visit(ls);
+    }
+
+    override void visit(LabelStatement ls)
+    {
+        assert(ls);
+        ls.loc.warning("TODO detect unused label `%s`", ls.toChars());
+        super.visit(ls);
+    }
+
+    override void visit(Import im)
+    {
+        assert(im);
+        // TODO: detect: `import io = std.io;`
+        // if (!im.isReferenced)
+        //     im.loc.warning("unused import `%s`", im.toChars());
+        foreach (const ad; im.aliasdecls)
+        {
+            if (!ad.isReferenced)
+            {
+                if (ad.protection.kind != Prot.Kind.public_)
+                    ad.loc.warning("unused %s imported alias `%s`", im.prot.kind.toChars(), ad.toChars());
+            }
+        }
+        super.visit(im);
+    }
+}
+
+static private immutable(char)* toChars(in Prot.Kind protKind) @safe pure nothrow @nogc
+{
+    final switch (protKind)
+    {
+    case Prot.Kind.undefined:
+        return "undefined";
+    case Prot.Kind.none:
+        return "none";
+    case Prot.Kind.private_:
+        return "private";
+    case Prot.Kind.package_:
+        return "package";
+    case Prot.Kind.protected_:
+        return "protected";
+    case Prot.Kind.public_:
+        return "public";
+    case Prot.Kind.export_:
+        return "export";
+    }
+}
+
+static private bool shouldWarnUnunsed(in Prot.Kind protKind) @safe pure nothrow @nogc
+{
+    final switch (protKind)
```

Comment:
You can group the cases.

Reply:
Ahh, thanks. This could also be detect in a diagnostics! ;)